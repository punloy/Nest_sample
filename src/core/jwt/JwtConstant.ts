export enum JwtExpiredTime {
    API_8_HOURS_SECONDS = 8 * 60 * 60
}
