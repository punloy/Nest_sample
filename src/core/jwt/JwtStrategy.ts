import { Injectable } from "@nestjs/common";
import { PassportStrategy } from "@nestjs/passport";
import config from "config";
import { ExtractJwt, Strategy } from "passport-jwt";
import { JwtInfo } from "../dto/JwtInfo";
import { JwtPayload } from "../dto/JwtPayload";

const jwk: any = config.get("jwk");

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy, "jwt") {
    constructor() {
        super({
            jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
            ignoreExpired: false,
            secretOrKey: jwk
        });
    }

    public async validate({ id, data, jti }: JwtPayload): Promise<JwtInfo> {
        return { id, data, jti };
    }
}
