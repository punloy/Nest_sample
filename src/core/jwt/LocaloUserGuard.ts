import { Injectable } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";

@Injectable()
export class LocalUserGuard extends AuthGuard("local-user") { }
