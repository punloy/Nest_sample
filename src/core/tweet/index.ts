export { User } from "./entities/User";
export { Tweet } from "./entities/Tweet";
export { Follower } from "./entities/Follower";