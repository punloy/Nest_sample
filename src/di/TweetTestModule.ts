import { DynamicModule, Module } from "@nestjs/common";
import { TypeOrmModuleOptions } from "@nestjs/typeorm";
import config from "config";
import { AppModule } from "./AppModule";
import { TweetDatabaseModule } from "./TweetDatabaseModule";

const tweetOptions: Partial<TypeOrmModuleOptions> = config.get("tweetDB");

@Module({})
export class TweetTestModule {
    public static register(): DynamicModule {
        return {
            module: TweetTestModule,
            imports: [
                TweetDatabaseModule.register(tweetOptions),
                AppModule
            ]
        };
    }
}
