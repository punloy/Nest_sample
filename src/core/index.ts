import * as tweet from "./tweet"

export {
    tweet
}