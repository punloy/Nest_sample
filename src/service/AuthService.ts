import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { tweet } from "../core/index";
import { Repository } from "typeorm";
import { encodeBase64 } from "../core/utils/encodeBase64";

@Injectable()
export class AuthService {
    constructor(
        @InjectRepository(tweet.User)
        private readonly userRepository: Repository<tweet.User>
    ) { }

    public async validateUser(name: string, password: string): Promise<Partial<tweet.User>> {
        const user = await this.userRepository.findOne({
            where: { name }
        });

        if (user && user.passwordHash === encodeBase64(password)) {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { passwordHash, ...result } = user;
            return result;
        }

        return null;
    }
}