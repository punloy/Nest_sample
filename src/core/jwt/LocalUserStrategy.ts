import { ForbiddenException, Injectable } from "@nestjs/common";
import { PassportStrategy } from "@nestjs/passport";
import { tweet } from "..";
import { Strategy } from "passport-local";
import { AuthService } from "../../service/AuthService";

@Injectable()
export class LocalUsertrategy extends PassportStrategy(Strategy, "local-user") {
    constructor(private readonly authService: AuthService) {
        super();
    }

    async validate(name: string, password: string): Promise<Partial<tweet.User>> {
        const account = await this.authService.validateUser(name, password);
        if (!account) {
            throw new ForbiddenException("Username Or Password Is Not Correct");
        }
        return account;
    }
}
