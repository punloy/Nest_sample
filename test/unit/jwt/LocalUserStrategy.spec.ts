import { Test } from "@nestjs/testing";
import { TypeOrmModule } from "@nestjs/typeorm";
import { tweet } from "../../../src/core";
import { LocalUsertrategy } from "../../../src/core/jwt/LocalUserStrategy";
import { TweetTestModule } from "../../../src/di/TweetTestModule";
import { AuthService } from "../../../src/service/AuthService";
import { createTestUser } from "../../../test/helper/TestTweetData";


describe("LocalCustomerAccountStrategy spec", () => {
    let localUsertrategy: LocalUsertrategy;

    beforeAll(async () => {
        const moduleRef = await Test.createTestingModule({
            imports: [
                TweetTestModule.register(),
                TypeOrmModule.forFeature(Object.values(tweet))
            ],
            providers: [
                AuthService,
                LocalUsertrategy
            ]
        }).compile();

        localUsertrategy = moduleRef.get<LocalUsertrategy>(LocalUsertrategy);

        await createTestUser({ userId: 1 });
    });

    it("should validate", async () => {
        const localUserStrategy = await localUsertrategy.validate("test user 1", "password");
        expect(localUserStrategy.id).toBe(1);
        expect(localUserStrategy.name).toBe("test user 1");
        //expect(localUserStrategy.passwordHash).toBe(encodeBase64("password"));
    });

    it("should not validate if account is invalid", async () => {
        await expect(localUsertrategy.validate("test2", "password"))
            .rejects.toThrowErrorMatchingSnapshot();
    });

});