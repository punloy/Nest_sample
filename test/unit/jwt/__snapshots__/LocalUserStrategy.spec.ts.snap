// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalCustomerAccountStrategy spec should not validate if account is invalid 1`] = `"Username Or Password Is Not Correct"`;
